src/Juvix/Compiler/Core/Info/FreeVarsInfo.hs

```diff
@@ -0,0 +1,49 @@
+module Juvix.Compiler.Core.Info.FreeVarsInfo where
+
+import Data.HashMap.Strict qualified as HashMap
+import Juvix.Compiler.Core.Extra
+import Juvix.Compiler.Core.Info qualified as Info
+import Juvix.Compiler.Core.Language
+
+newtype FreeVarsInfo = FreeVarsInfo
+  { -- map free variables to the number of their occurrences
+    _infoFreeVars :: HashMap Index Int
+  }
+
+instance IsInfo FreeVarsInfo
+
+kFreeVarsInfo :: Key FreeVarsInfo
+kFreeVarsInfo = Proxy
+
+makeLenses ''FreeVarsInfo
+
+computeFreeVarsInfo :: Node -> Node
+computeFreeVarsInfo = umap go
+  where
+    go :: Node -> Node
+    go node = case node of
+      NVar Var {..} ->
+        mkVar (Info.insert fvi _varInfo) _varIndex
+        where
+          fvi = FreeVarsInfo (HashMap.singleton _varIndex 1)
+      _ ->
+        modifyInfo (Info.insert fvi) node
+        where
+          fvi =
+            FreeVarsInfo $
+              foldr
+                ( \NodeChild {..} acc ->
+                    HashMap.unionWith (+) acc $
+                      HashMap.mapKeys (\idx -> idx - _childBindersNum) $
```

Comment:
Have you considered using `Map` instead of `HashMap`?
We are doing lots of `unionWith` and `mapKeys`. [`Map.unionWith`](https://hackage.haskell.org/package/containers-0.6.7/docs/Data-Map-Strict.html#v:unionWith) is slightly more efficient than [`HashMap.unionWith`](https://hackage.haskell.org/package/containers-0.6.7/docs/Data-Map-Strict.html#v:unionWith). We could also use [`Map.mapKeysMonotonic`](https://hackage.haskell.org/package/containers-0.6.7/docs/Data-Map-Strict.html#v:mapKeysMonotonic).